Add scaffolder layouts to next scaffolder page

dagda1 opened the pull request:
## Hey, I just made a Pull Request!

When I created #13112, I only added the `ScaffolderLayouts` to the existing `ScaffolderPage`.

We have been switching over to the `NextScaffolderPage`, and our clients have been using the `ScaffolderLayouts`.

This PR adds them to the `NextScaffolderPage` component hierarchy.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

benjdlambert commented on plugins/scaffolder/src/next/Router/Router.tsx:
```diff
@@ -95,7 +106,10 @@ export const Router = (props: PropsWithChildren<NextRouterProps>) => {
-              FormProps={props.FormProps}
+              FormProps={{
+                ...props.FormProps,
+                layouts: customLayouts,
```
I think that these should just be on the `TemplateWizardPage` as they're not props that are on the `Form` especially, they're props that we convert into some concept known on the `Form` itself. They're pretty similar the `customFieldExtensions`.

dagda1 replied:
@benjdlambert agreed

benjdlambert commented on plugins/scaffolder/src/next/types.ts:
```diff
@@ -15,6 +15,25 @@
 import type { FormProps as SchemaFormProps } from '@rjsf/core-v5';
 
+/**
+ * The field template from \@rjsf/core which is a react component that gets passed \@rjsf/core field related props.
```
Think you can backticks instead of escaping here to remove the warning and make it look nice in the docs.

dagda1 replied:
done

benjdlambert commented on plugins/scaffolder/src/next/TemplateWizardPage/Stepper/useTemplateSchema.ts:
```diff
@@ -44,25 +48,40 @@ export const useTemplateSchema = (
-                : true;
-            },
+    .map(step => {
+      const objectFieldTemplate = step.uiSchema['ui:ObjectFieldTemplate'] as
```
Not sure if it makes sense to do this stuff in this hook. I'm thinking it makes more sense to do it somewhere in the `Stepper` component, or we write a hook to do this separately instead and call that hook in the stepper with the output of this hook maybe.

dagda1 replied:
I've created a `useTransformSchemaToProps` function that does this now.

benjdlambert commented on plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx:
```diff
@@ -37,7 +37,7 @@ import { ReviewState } from './ReviewState';
 import { selectedTemplateRouteRef } from '../../../routes';
 import { useFormData } from './useFormData';
-import { FormProps } from '../../types';
+import type { FormProps as FormFields, NextLayoutOptions } from '../../types';
```
Why the rename?

dagda1 replied:
I honestly cannot remember...removed.

benjdlambert reviewed (changes requested):
Some higher level comments here about where to place this logic, not sure we want to overload the meaning of that hook and what it does.

benjdlambert commented:
I added a todo in the #15399 to move the collection logic for these layouts to `-react` like the `useCustomFieldExtensions` hook that exists. That PR should land today, so it probably makes sense to rebase on top of that.

dagda1 commented:
let's keep this open

benjdlambert commented:
This should be good to rebase again on top of the `-react` work. There's a bit that's changed, but you should be able to follow the changesets for the changes. I only moved the `next` components out into the `-react` package as it didn't seem worth it to create a deprecation path for those when they're going to become redundant soon.

dagda1 commented:
> Some higher level comments here about where to place this logic, not sure we want to overload the meaning of that hook and what it does.



> This should be good to rebase again on top of the `-react` work. There's a bit that's changed, but you should be able to follow the changesets for the changes. I only moved the `next` components out into the `-react` package as it didn't seem worth it to create a deprecation path for those when they're going to become redundant soon.

I've rebased and updated.

benjdlambert commented on plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:
```diff
@@ -73,7 +74,7 @@ const Form = withTheme(require('@rjsf/material-ui-v5').Theme);
  * @alpha
  */
-export const Stepper = (props: StepperProps) => {
+export const Stepper = ({ layouts = [], ...props }: StepperProps) => {
```
Let's destructure inside the component like before :)

benjdlambert commented on plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:
```diff
@@ -141,6 +142,8 @@ export const Stepper = (props: StepperProps) => {
     setFormState(current => ({ ...current, ...formData }));
   };
 
+  const currentStep = useTransformSchemaToProps(steps[activeStep], layouts);
```
Think that we should make the second argument an object and have layouts be optional so that we can evolve the api over time.

benjdlambert commented on plugins/scaffolder/src/components/MultistepJsonForm/schema.ts:
```diff
@@ -120,7 +120,8 @@ export function transformSchemaToProps(
 
     if (Layout) {
-      uiSchema['ui:ObjectFieldTemplate'] = Layout;
+      uiSchema['ui:ObjectFieldTemplate'] =
```
Hmm this cast is unfortunate. I guess it's because it's using the `uiSchema` type from the later versions of `@rjsf/*` right?

dagda1 replied:
@benjdlambert pretty much, in the `@rjsf/core-v5` schema, the `ObjectFieldTemplate` uses `ComponentType`

```ts
ObjectFieldTemplate: React.ComponentType<ObjectFieldTemplateProps<T, S, F>>;
```

But in `@rjsf/core-3.2.0`, it uses `StatelessComponent`

```ts
'ui:ObjectFieldTemplate'?: React.StatelessComponent<ObjectFieldTemplateProps>;
```

`ComponentType` is a union of `ComponentClass` and `StatelessComponent` so typescript will not do the conversion because the 3.2.0 type is only for `StatelessComponent` and not `ClassComponent`.

benjdlambert commented on plugins/scaffolder/src/next/Router/Router.tsx:
```diff
@@ -82,7 +85,10 @@ export const Router = (props: PropsWithChildren<NextRouterProps>) => {
-              FormProps={props.FormProps}
+              layouts={customLayouts}
+              FormProps={{
+                ...props.FormProps,
```
Don't need the spreading here?

dagda1 replied:
ha!, good spot

benjdlambert commented on plugins/scaffolder/api-report.md:
```diff
@@ -56,10 +58,8 @@ import { z } from 'zod';
-export function createScaffolderLayout<TInputProps = unknown>(
-  options: LayoutOptions,
-): Extension<LayoutComponent<TInputProps>>;
+// @public @deprecated (undocumented)
```
We should document these deprecations in the changesets and suggest how to fix them :)

benjdlambert replied:
Maybe a seperate changeset entirely where you can provide examples similar to the ones for #15399

benjdlambert reviewed (changes requested):
Nice work! some comments! sorry it's taken a while to get round to this again!

benjdlambert commented on .changeset/friendly-scissors-shop.md:
````diff
@@ -0,0 +1,21 @@
+  LayoutTemplate
+  ```
+
+- The following `/alpha` types have removed from this package and moved to the `@backstage/plugin-scaffolder-react/alpha` package
````
Is this true?

dagda1 replied:
It is.  Why do you ask?

benjdlambert replied:
Looks like at least `createScaffolderLayout` is `@public` so not available in `/alpha` I think we should just remove the second part of this as they're all available in `@backstage/plugin-scaffolder-react` now right?

benjdlambert replied:
https://github.com/backstage/backstage/pull/15414/files#diff-30803124f64dd6757ee54c8081a2bb05c00dbbdb1f90fa386203a8c5ddceb9e7R68-R72

dagda1 replied:
sorry, my bad.  Should i just remove the second changeset?

benjdlambert replied:
I didn't mean didn't delete the changeset. 😅  Just the second part 👍

dagda1 replied:
> I didn't mean didn't delete the changeset. 😅 Just the second part 👍

lol, reinstated.  I've had too much coffee or not enough

benjdlambert commented:
Thanks for the patience on this one! :pray:
